src/Common/src/CoreLib/System/Memory.cs

```diff
@@ -353,13 +353,26 @@ public unsafe MemoryHandle Pin()
             }
             else if (_object is T[] array)
             {
-                GCHandle handle = _length < 0 ? default : GCHandle.Alloc(array, GCHandleType.Pinned);
+                // Array is already pre-pinned
+                if (_length < 0)
+                {
 #if FEATURE_PORTABLE_SPAN
-                void* pointer = Unsafe.Add<T>((void*)handle.AddrOfPinnedObject(), _index);
```

Comment:
I want to highlight a fix for a bug here where handle could be null (for portable span). Please review.

**Tests from this PR caught the bug:**
https://mc.dot.net/#/user/ahsonkhan/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/f4b9ba1fd81623c838cc27647936c3569e83aa22/workItem/System.Memory.Tests/analysis/xunit/System.SpanTests.MemoryMarshalTests~2FCreateFromPinnedArrayVerifyPinning
System.SpanTests.MemoryMarshalTests/CreateFromPinnedArrayVerifyPinning
```text
System.InvalidOperationException : Handle is not initialized.
Stack Trace :
   at System.Runtime.InteropServices.GCHandle.AddrOfPinnedObject()
   at System.Memory`1.Pin() in D:\j\workspace\windows-TGrou---2a8f9c29\src\Common\src\CoreLib\System\Memory.cs:line 358
   at System.SpanTests.MemoryMarshalTests.CreateFromPinnedArrayVerifyPinning() in D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Memory\tests\MemoryMarshal\CreateFromPinnedArray.cs:line 140
```

cc @stephentoub, @KrzysztofCwalina, @jkotas